### PR TITLE
20.0 paypal paylater messages

### PR DESCRIPTION
**Pull request by ProxiBlue**
### Description (*)
PayPal US has a new feature called Pay Later (or Pay in 4)
This ads in the required messages to inform user that can Pay in 4 instalments, if they want to do so.

ref: https://www.paypal.com/us/for-you/pay-in-4

### Fixed Issues (if relevant)
1. Fixes OpenMage/magento-lts #1435

### Manual testing scenarios (*)

1. In admin setup Paypal Express, must use a US based account (vital)
2. Set Paypal express buttons for Product and Cart to be enabled
2. In admin set Merchant location to US, enable Pay Later messages, set PayPal SDK client id (ref: https://developer.paypal.com/docs/business/pay-later/integrate/)

image

3. In frontend, browse to any product. You will see a message generated that shows the price of the product viewed split PRICE / 4

image

4. Add item to cart (and more items to cart if you so wish)
5. In cart you will get same messaging, based on cart grand total / 4

image

6. Proceed to checkout. You will get same messaging on payment selection

image

7. In paypal wallet, user wil get ability to do the Pay 4 method. It will ONLY work for USA based paypal accounts/merchants

### Questions or comments

First attempt to implement this. Looking for a peer review / suggestions on doing something in code better.
Not sure how to enact unit testing for this.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by Flyingmana — DISMISSED** [review]
Code looks ok

**Comment by ProxiBlue** [comment]
Just noticed this contains unwanted file mode changes on media files

image

I am removing

Might be a need for an automated test, any edits/changes in media folder ?

**Comment by pauldpauld** [comment]
Just noting PayPal Australia have announced;
In an announcement on Wednesday, the Nasdaq-listed giant said its BNPL solution will be available to its 9 million Australian customers in early June (2021), ready for use by the end of financial year sales.

**Comment by ProxiBlue** [comment]
@pauldpauld 

Yep, they have a tough market to break into, with 3 established services already in operation, and doing well.
That said, I will adjust this PR to allow Australia in country selections, as it was locked to USA only for config availability

**Comment by ProxiBlue** [comment]
FWIW, PayPal has offered my existing clients lowered rates to display this messages, in a fairly aggressive push to market in US.
WIll be interesting to see what incentives they will offer AU side.

**Comment by ProxiBlue** [comment]
Adjusted to allow Australia, and general improvement on actual way it was implemented on cart and product pages

File: skin/frontend/rwd/default/css/styles.css [hunk 1]
```diff
@@ -1072,6 +1072,20 @@ a.button:hover {
     float: none;
   }
 }
+
+/* -------------------------------------------- *
```
Comment (fballiano) [inline skin/frontend/rwd/default/css/styles.css:1076]: we should update also the scssf files or this change will be lost when css will be regenerated.

**Comment by matteotestoni** [comment]
Hi, paypal pay later is now available in italy and other countries. When you merge this patch?

**Comment by fballiano** [comment]
@matteotestoni did you test this PR? if so can you add an "approved" review? thanks!

File: app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?php
+/**
+ * Paypal Pay Later
+ *
+ * https://www.paypal.com/us/for-you/pay-in-4
+ * This empty class exists to facilitate easy 3rd party extending Cart Messaging functionality
+ *
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php:7]:
```suggestion
/**
 * OpenMage
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * https://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@magento.com so we can send you a copy immediately.
 *
 * @category   Mage
 * @package    Mage_Paypal
 * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
 * @license    https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

/**
 * @category   Mage
 * @package    Mage_Paypal
 * @author     Magento Core Team <core@magentocommerce.com>
 */
```

File: app/code/core/Mage/Paypal/Block/Paylater/Message.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
+<?php
+
+/**
+ * Paypal Pay Later
+ *
+ * Available to specific list of merchant countries only
+ *
+ */
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Message.php:8]:
```suggestion
/**
 * OpenMage
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * https://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@magento.com so we can send you a copy immediately.
 *
 * @category   Mage
 * @package    Mage_Paypal
 * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
 * @license    https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

/**
 * PayPal Standard payment "form"
 *
 * @category   Mage
 * @package    Mage_Paypal
 * @author     Magento Core Team <core@magentocommerce.com>
 */
```

File: app/code/core/Mage/Paypal/Block/Paylater/Onepage/Message.php [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?php
+/**
+ * Paypal Pay Later
+ *
+ * https://www.paypal.com/us/for-you/pay-in-4
+ * This empty class exists to facilitate easy 3rd party extending Onepage Messaging functionality
+ *
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Onepage/Message.php:7]:
```suggestion
/**
 * OpenMage
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * https://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@magento.com so we can send you a copy immediately.
 *
 * @category   Mage
 * @package    Mage_Paypal
 * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
 * @license    https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

/**
 * PayPal Standard payment "form"
 *
 * @category   Mage
 * @package    Mage_Paypal
 * @author     Magento Core Team <core@magentocommerce.com>
 */
```

File: app/code/core/Mage/Paypal/Block/Paylater/Product/Message.php [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?php
+/**
+ * Paypal Pay Later
+ *
+ * https://www.paypal.com/us/for-you/pay-in-4
+ * This empty class exists to facilitate easy 3rd party extending Product Messaging functionality
+ *
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Product/Message.php:7]:
```suggestion
/**
 * OpenMage
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * https://opensource.org/licenses/osl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@magento.com so we can send you a copy immediately.
 *
 * @category   Mage
 * @package    Mage_Paypal
 * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
 * @license    https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

/**
 * PayPal Standard payment "form"
 *
 * @category   Mage
 * @package    Mage_Paypal
 * @author     Magento Core Team <core@magentocommerce.com>
```

**Comment by ADDISON74** [comment]
From what I can see in the source code this PR is only valid for the USA and Australia. If it is also valid for other countries, changes must be made, in the Backend only US, AU are listed as countries codes.

**Comment by matteotestoni** [comment]
image

**Comment by matteotestoni** [comment]
maybe a multiselect where user choose countries where to activate pay later.
Would be a good new feature

File: app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ * Paypal Pay Later
+ *
+ * https://www.paypal.com/us/for-you/pay-in-4
+ * This empty class exists to facilitate easy 3rd party extending Cart Messaging functionality
+ * 
+ * @category   Mage
+ * @package    Mage_Paypal
+ * @author     Magento Core Team <core@magentocommerce.com>
+ */
+ */
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php:32]:
```suggestion
```

**Comment by ProxiBlue** [comment]
Hello, FYI, I don;t currently have any capacity to fix any issues, update any code to effect this 2 year old MR
Feel free to fork, and update accordingly.

Maybe sometime in december.

**Comment by empiricompany** [comment]
I have written a 3rd party extension if you are interested: https://github.com/empiricompany/openmage-paypal-pay-later-banner-info

File: app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/osl-3.0.php
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@magento.com so we can send you a copy immediately.
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php:14]:
```suggestion
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available at https://opensource.org/license/osl-3-0-php
```

File: app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
+<?php
+/**
+/**
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Cart/Message.php:3]:
```suggestion
/**
```

File: app/code/core/Mage/Paypal/Block/Paylater/Message.php [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/osl-3.0.php
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@magento.com so we can send you a copy immediately.
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Message.php:13]:
```suggestion
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available at https://opensource.org/license/osl-3-0-php
```

File: app/code/core/Mage/Paypal/Block/Paylater/Product/Message.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Product/Message.php:2]:
```suggestion
```

File: app/code/core/Mage/Paypal/Block/Paylater/Product/Message.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/osl-3.0.php
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@magento.com so we can send you a copy immediately.
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Product/Message.php:13]:
```suggestion
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available at https://opensource.org/license/osl-3-0-php
```

File: app/code/core/Mage/Paypal/Block/Paylater/Message.php [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+     * Valid countries that the feature will work on.
+     * Any other country set will disable feature
+     *
+     * It is likely only ever going to be USA
+     *
+     * @var array
+     */
+    protected $_validCountryCodes = [Mage_Paypal_Helper_Data::US_COUNTRY, self::AU_COUNTRY];
```
Comment (fballiano) [inline app/code/core/Mage/Paypal/Block/Paylater/Message.php:39]: this is now outdated since more countries are available

**Comment by fballiano** [comment]
Since this PR is now outdated because of https://github.com/OpenMage/magento-lts/pull/1437#discussion_r1157078079 I think we should add @empiricompany's module to our recommended modules (it has many more features) and close this PR.

**Comment by fballiano** [comment]
@empiricompany I've added your module on our "readthedocs" and on https://github.com/fballiano/awesome-openmage

**Review by matteotestoni — APPROVED** [review]

**Comment by ADDISON74** [comment]
I am closing this PR. There is already a solution to the reported issue here https://github.com/OpenMage/magento-lts/pull/1437#issuecomment-1418755307. Also, the author is not able to fix any requests or code issues.

**Pull request closed without merging**